### ISSUE TITLE
Using a variable named buildcmd on easyconfigs shouldn't be fatal

### ISSUE THREAD

**Issue by gppezzi** [easybuilders/easybuild-framework#2912]
This [vmd easyconfig](https://github.com/eth-cscs/production/blob/d542cafe39008cdae4913a56370426ba5b855220/easybuild/easyconfigs/v/VMD/VMD-1.9.3-egl.eb) was working happily until we upgraded to EB 3.9.1.

With EB 3.9.1 we get a fatal error:
```
ERROR: Failed to process easyconfig /users/perettig/VMD-1.9.3-egl.eb: You may have some typos in your easyconfig file: buildcmd -> build_cmd
``` 
Maybe it would be more useful to have a warning and proceed with the build?

**Comment by boegel** [comment]
This issue is caused by the `build_cmd` easyconfig parameter that was added to the generic `ConfigureMake` easyblock via https://github.com/easybuilders/easybuild-easyblocks/pull/1628.

In this case, `buildcmd` was basically just being used as a local variable, the intention was never to use the newly `build_cmd` introduced easyconfig parameter at all (in fact, in this context it's entirely meaningless, since `CmdCp` doesn't use `build_cmd` at all).

With this in mind, it probably makes sense to let the typo detection issue a clear warning rather than producing an error, since in this case it shouldn't be fatal at all...

I'm a bit torn on whether making that change is a good idea though, since this is actually a sign of a bigger issue where innocent local variables that are not known easyconfig parameters in EasyBuild version X (and thus are basically totally ignored by EasyBuild itself, they only serve a purpose while the easyconfig file is parsed) can suddenly be assigned meaning in EasyBuild version X+1 where additional easyconfig parameters were defined. The typo detection causes trouble here is just bad luck, but it could've been "worse" in theory.

I'm not sure how we can avoid the bigger issue from occurring going forward, other than introducing a requirement that local variables follow a particular naming scheme (e.g. `local_xxx` rather than `xxx`), and make EasyBuild complain if any local variables that don't follow this scheme are being used. That would be a fairly painful change to make, since it would break any easyconfig file that uses local variables (although it may be feasible to auto-fix those).

**Comment by vanzod** [comment]
@boegel This is definitely a potential source of issues. I agree on your last paragraph.

**Comment by victorusu** [comment]
IMHO, @boegel, I agree with the idea of `local_xxx` and I think it should be a guideline, but I also have sympathies for an imposition by the framework to force local variable names.

I think it is best to keep it as a fatal error. It is a good practice to avoid having similar variables names to the ones of the easyconfig API. In the long run, a similar variable name would hurt even more.  

IMHO, Easybuild prints a lot of lines on the console already. Force it produce another warning will pollute even more the console.

**Comment by akesandgren** [comment]
Painful as it may be enforcing local_xxxx is a good way forward in my opinion.
But probably a thing for 4.0 since it may break local easyconfigs.

**Comment by ocaisa** [comment]
Agree in general, I think failing hard is ok but we can improve the error message to explain how to get around the problem if the usage was explicitly intended to be a local variable (`if this variable is meant to be local please rename to local_xxxx`). I can open a PR for that.

**Comment by ocaisa** [comment]
@gppezzi I don't get the same error as you with this easyconfig:
```
alanc@~/EasyBuild_Git/easybuild-framework(develop)$ eb -x VMD-1.9.3-egl.eb 
== temporary log file in case of crash /tmp/eb-uktsbf/easybuild-Sq8IVB.log
ERROR: Failed to process easyconfig /home/alanc/EasyBuild_Git/easybuild-framework/VMD-1.9.3-egl.eb: You may have some typos in your easyconfig file: buildcmd -> build_cmd
```
What does your command line look like to produce the error in your original post?

**Comment by gppezzi** [comment]
@ocaisa my bad, I copied the wrong error message (probably after having tweaked that eb file): I actually get the same error message and I've updated the issue accordingly.

Sorry for that.

**Comment by gppezzi** [comment]
I also agree that changing the behaviour could do more harm than good, so maybe just improving the error message can be a good compromise for avoiding to hit the same misleading error message in the future.

**Comment by boegel** [comment]
@gppezzi We're taking some measures to counter this situation in EasyBuild 4.0, which were implemented in #2938.

Executive summary:

* the type error stays, pretty much unchanged
* there's a recommended naming scheme for local variables in easyconfigs (`_*`, `local_*` or single-letter variables, the latter because of list comprehensions)
* `eb` will issue a warning if any non-conforming local variables were found
* the warning can be promoted to an error in case you want to be more strict

Documentation update on this coming soon (see https://boegel-eb.readthedocs.io/en/local_vars/Easyconfig-files-local-variables.html for an early preview).

**Comment by boegel** [comment]
Closing this now, since we've done the necessary work (currently in `4.x` branch) to avoid that local variables could become problematic later, see more details at https://easybuild.readthedocs.io/en/4.x/Easyconfig-files-local-variables.html
